### PR TITLE
[ADD] mail_restrict_follower_selection

### PR DESCRIPTION
**Pull request by hbrunn**
# Restrict follower selection

This module was written to allow you to restrict the selection of possible followers. For example, if you use the social ERP functions only internally, it makes sense to filter possible followers for being employees. Otherwise, you'll get a quite crowded list of partners to choose from.
# Configuration

To configure this module, you need to go to `System parameters` and adjust `mail_restrict_follower_selection.domain` as you see fit. This restricts followers globally, if you want to restrict only the followers for a certain record type (or have different restrictions for different record types), create a parameter `mail_restrict_follower_selection.domain.$your_model`.

Note: This module won't change existing followers!

### CODE REVIEW

**Comment by dreispt** [comment]
Do you consider also adding a filter to the `mail.compose.message` wizard?
This can be is opened when adding a new chatter message, and allows to pick Partners that will also be added as Followers.

I have a use case where In would like like to limit possible contacts to only those that have an email address.

**Comment by hbrunn** [comment]
sounds good. does it make sense to have this in a different parameter or use the same one?

**Comment by dreispt** [comment]
That's a good question. I say it's far from overwhelming to have two parameters, and it could be nice to have that flexibility.

File: mail_restrict_follower_selection/README.rst [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    :alt: License: AGPL-3
+Restrict follower selection
+===========================
+
+This module was written to allow you to restrict the selection of possible followers. For example, if you use the social ERP functions only internally, it makes sense to filter possible followers for being employees. Otherwise, you'll get a quite crowded list of partners to choose from.
+
+Configuration
+=============
+
+To configure this module, you need to go to `System parameters` and adjust `mail_restrict_follower_selection.domain` as you see fit. This restricts followers globally, if you want to restrict only the followers for a certain record type (or have different restrictions for different record types), create a parameter `mail_restrict_follower_selection.domain.$your_model`.
```
Comment (pedrobaeza) [inline mail_restrict_follower_selection/README.rst:11]: Please put an example of a valid domain.

File: mail_restrict_follower_selection/__init__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    OpenERP, Open Source Management Solution
```
Comment (pedrobaeza) [inline mail_restrict_follower_selection/__init__.py]: You can remove this line or maybe adjust to new brand

File: mail_restrict_follower_selection/models/__init__.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    OpenERP, Open Source Management Solution
```
Comment (pedrobaeza) [inline mail_restrict_follower_selection/models/__init__.py]: The same

File: mail_restrict_follower_selection/README.rst (hunk 1 above)
Comment (dreispt) [inline mail_restrict_follower_selection/README.rst:11]: @hbrunn The default parameter created has a key `mail_restrict_follower_selection.domain`. Does this work as a global value?

File: mail_restrict_follower_selection/models/mail_wizard_invite.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+##############################################################################
+from lxml import etree
+from openerp import models, api
+
+
+class MailWizardInvite(models.TransientModel):
+    _inherit = 'mail.wizard.invite'
+
+    @api.model
+    def mail_restrict_follower_selection_get_domain(self):
```
Comment (dreispt) [inline mail_restrict_follower_selection/models/mail_wizard_invite.py]: Add underscore prefix to make it private?

File: mail_restrict_follower_selection/README.rst [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+Note: This module won't change existing followers!
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/crm/issues>`_.
```
Comment (adrienpeiffer) [inline mail_restrict_follower_selection/README.rst]: Replace `crm` by `social`

File: mail_restrict_follower_selection/README.rst [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/crm/issues/new?body=module:%20mail_restrict_follower_selection%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Comment (adrienpeiffer) [inline mail_restrict_follower_selection/README.rst]: Same here

File: mail_restrict_follower_selection/__openerp__.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+{
+    "name": "Restrict follower selection",
+    "version": "1.0",
```
Comment (adrienpeiffer) [inline mail_restrict_follower_selection/__openerp__.py]: 8.0.1.0.0

File: mail_restrict_follower_selection/__openerp__.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    "author": "Therp BV",
```
Comment (adrienpeiffer) [inline mail_restrict_follower_selection/__openerp__.py]: Add OCA as author

File: mail_restrict_follower_selection/README.rst (hunk 1 above)
Comment (hbrunn) [inline mail_restrict_follower_selection/README.rst:11]: @dreispt yes, that's what I want to express in "...adjust `mail_restrict_follower_selection.domain` as you see fit. This restricts followers globally" - unclear?

**Comment by hbrunn** [comment]
@dreispt @adrienpeiffer @pedrobaeza thanks for your reviews, all suggestions taken & commits squashed

File: mail_restrict_follower_selection/README.rst (hunk 1 above)
Comment (pedrobaeza) [inline mail_restrict_follower_selection/README.rst:11]: For me it's OK

**Comment by pedrobaeza** [comment]
:+1:

**Comment by rafaelbn** [comment]
Hi @hbrunn , tested in runbot.

By default runbot has:
- mail_restrict_follower_selection.domain      [('employee', '=', True)]
- runbot has 25 employee

When I try to add a follower to any object (employee, sale order, task, ...) I don't have any choice. Is this default parameter correct? It looks is not working.

**Comment by dreispt** [comment]
Well, the module does not depend on `hr`, so having a default setting based on employees looks awkward...

**Comment by hbrunn** [comment]
@rafaelbn @dreispt it's a field from base: https://github.com/OCA/OCB/blob/8.0/openerp/addons/base/res/res_partner.py#L261 - and you have to set that of course to find anything.

But considering the confusion, I changed the default to customer, that should give everyone some results.

**Comment by pedrobaeza** [comment]
Employee option is not that bad (to avoid leaks to outside) if it's explained, because people don't know about that field in general.

**Comment by hbrunn** [comment]
When I looked into that, I noticed that the standard UI doesn't allow to set this field, so I don't think it's a good choice for a default. And it's just a default anyways, I don't really think end users install this module, but integrators use it to set up whatever restrictions they see fit.

**Comment by pedrobaeza** [comment]
Yeah, that's right

**Comment by dreispt** [comment]
:+1: !

**Comment by pedrobaeza** [comment]
@rafaelbn, give us your approval also and I'll merge

**Comment by yajo** [comment]
Maybe adding to default & instructions `('opt_out', '=', False)` would be a good idea.

**Comment by yajo** [comment]
Also add to instructions that `("user_ids","!=",False)` will enable only actual users in your database.

**Comment by yajo** [comment]
I noticed that when you install the first module after you installed `mail_restrict_follower_selection` it leads the user to this strange page that shows the default value for its setting. (In runbot the url is http://3115521-1-2053bb.runbot1.odoo-community.org/web?db=3115521-1-2053bb-all#id=5&view_type=form&model=ir.config_parameter&action=110&active_id=268):

captura de pantalla de 2015-09-21 19-25-39

It's not something critical though.

**Comment by rafaelbn** [comment]
Hi @hbrunn ,

With Sales, Project, HR installed I tried to restrict only for employees, globaly, but I can't. Maybe I'm missing something you say here https://github.com/OCA/social/pull/1#issuecomment-137678952

My param is this one again (with HR installed):

```
Key
    mail_restrict_follower_selection.domain
Value
    [('employee', '=', True)] 
```

I expect only employee in follower bit it doesn't works.

I have also tried to play with more parameters with 

`mail_restrict_follower_selection.domain.$your_model`

in project.task, sale.orders, etc.. but it doesn't works. For sure I don't understand something.

Thanks

**Comment by hbrunn** [comment]
as written in https://github.com/OCA/social/pull/1#issuecomment-137678952, the employee flag does nothing. If you want to use it, you need to set it yourself

**Comment by yajo** [comment]
What do you (and others) think about https://github.com/OCA/social/pull/1#issuecomment-142038065?

**Comment by rafaelbn** [comment]
Hi @hbrunn could you check comments of @Yajo . They are logic and I think could be considered. After this we can finally merge it! Thanks!

**Comment by hbrunn** [comment]
I'm happy to merge any PRs into this branch, but extra work on my side for a default is expected too much

**Comment by rafaelbn** [comment]
:+1: on my site!

**Comment by yajo** [comment]
OK then let's merge and I'll add it later. :+1:

**Comment by rafaelbn** [comment]
Hi @Yajo please take care of pending PR about this one.

**Pull request merged**